ion/src/test/java/com/fasterxml/jackson/dataformat/ion/IonFactoryTest.java

```diff
@@ -0,0 +1,82 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.StringReader;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Parameterized.class)
+public class IonFactoryTest {
+
+    // 4-byte Ion 1.0 IVM followed by int 0.
+    private static final byte[] BINARY_INT_0 = new byte[] {(byte) 0xE0, 0x01, 0x00, (byte) 0xEA, 0x20};
+    private static final String TEXT_INT_0 = "0";
+
+    private enum InputSource {
+        BYTE_ARRAY() {
+            @Override
+            IonParser newParser(IonFactory factory) throws IOException {
+                return (IonParser) factory.createParser(BINARY_INT_0);
+            }
+        },
+        CHAR_ARRAY() {
+            @Override
+            IonParser newParser(IonFactory factory) throws IOException {
+                return (IonParser) factory.createParser(TEXT_INT_0.toCharArray());
+            }
+        },
+        READER() {
+            @Override
+            IonParser newParser(IonFactory factory) throws IOException {
+                return (IonParser) factory.createParser(new StringReader(TEXT_INT_0));
+            }
+        },
+        INPUT_STREAM() {
+            @Override
+            IonParser newParser(IonFactory factory) throws IOException {
+                return (IonParser) factory.createParser(new ByteArrayInputStream(BINARY_INT_0));
+            }
+        },
+        ION_READER() {
+            @Override
+            IonParser newParser(IonFactory factory) {
+                return factory.createParser(factory.getIonSystem().newReader(BINARY_INT_0));
+            }
+        },
+        ION_VALUE() {
+            @Override
+            IonParser newParser(IonFactory factory) {
+                return factory.createParser(factory.getIonSystem().newInt(0));
+            }
+        };
+
+        abstract IonParser newParser(IonFactory factory) throws IOException;
+    }
+
+    @Parameterized.Parameters(name = "{0}")
+    public static InputSource[] parameters() {
+        return InputSource.values();
+    }
+
+    @Parameterized.Parameter
+    public InputSource inputSource;
+
+    @Test
+    public void testIonParserIsResourceManaged() throws IOException {
+        IonFactory factory = new IonFactory();
+        IonParser parser = inputSource.newParser(factory);
+        // When the user provides an IonReader, it is not resource-managed, meaning that the user retains the
+        // responsibility to close it. In all other cases, the IonReader is created internally, is resource-managed,
+        // and is closed automatically in IonParser.close().
+        assertEquals(inputSource != InputSource.ION_READER, parser._ioContext.isResourceManaged());
+        assertTrue(IonReader.class.isAssignableFrom(parser._ioContext.contentReference().getRawContent().getClass()));
+        parser.close();
+    }
```

Comment:
```suggestion
    @Test
    public void byteArrayIsManaged() throws Throwable {
        assertResourceManaged(true, parser(f -> f.createParser(BINARY_INT_0)));
    }

    @Test
    public void charArrayIsManaged() throws Throwable {
        assertResourceManaged(true, parser(f -> f.createParser(TEXT_INT_0.toCharArray())));
    }

    @Test
    public void readerIsManaged() throws Throwable {
        assertResourceManaged(true, parser(f -> f.createParser(new StringReader(TEXT_INT_0))));
    }

    @Test
    public void inputStreamIsManaged() throws Throwable {
        assertResourceManaged(true, parser(f -> f.createParser(new ByteArrayInputStream(BINARY_INT_0))));
    }

    @Test
    public void ionValueIsManaged() throws Throwable {
        assertResourceManaged(true, parser(f -> f.createParser(f.getIonSystem().newInt(0))));
    }

    @Test
    public void ionReaderIsNotManaged() throws Throwable {
        // When the user provides an IonReader, it is not resource-managed, meaning that the user retains the
        // responsibility to close it. In all other cases, the IonReader is created internally, is resource-managed,
        // and is closed automatically in IonParser.close().
        assertResourceManaged(false, parser(f -> f.createParser(f.getIonSystem().newReader(BINARY_INT_0))));
    }

    private void assertResourceManaged(boolean expectResourceManaged, ThrowingSupplier<IonParser> supplier)
            throws Throwable {
        IonParser parser = supplier.get();
        Assertions.assertEquals(expectResourceManaged, parser._ioContext.isResourceManaged());
        Assertions.assertTrue(IonReader.class.isAssignableFrom(parser._ioContext.contentReference().getRawContent().getClass()));
        parser.close();
    }

    private interface ThrowingFunction<T,R> {
        R apply(T t) throws Throwable;
    }

    private static ThrowingSupplier<IonParser> parser(ThrowingFunction<IonFactory, JsonParser> f) {
        return () -> (IonParser) f.apply(new IonFactory());
    }
```

Reply:
We can drop a dozen lines and the parameterization at the same time without increasing the maintenance burden of the test cases imo. `ThrowingSupplier` and `Assertions` here come from JUnit 5 (in use by jackson-core so it ought to be fine to bring in). If you want to stick on JUnit 4.13 then `ThrowingSupplier` is another 3 lines. 

If we really want parameterized tests then we can use Java 8's functional features to drop 20 lines from the enum and make it more readable:

```java
    enum InputSource {
        BYTE_ARRAY(true, parser(f -> f.createParser(BINARY_INT_0))),
        CHAR_ARRAY(true, parser(f -> f.createParser(TEXT_INT_0.toCharArray()))),
        READER(true, parser(f -> f.createParser(new StringReader(TEXT_INT_0)))),
        INPUT_STREAM(true, parser(f -> f.createParser(new ByteArrayInputStream(BINARY_INT_0)))),
        ION_VALUE(true, parser(f -> f.createParser(f.getIonSystem().newInt(0)))),

        ION_READER(false, parser(f -> f.createParser(f.getIonSystem().newReader(BINARY_INT_0)))),
        ;

        private final boolean isResourceManaged;
        private final ThrowingSupplier<IonParser> supplier;

        InputSource(boolean isResourceManaged, ThrowingSupplier<IonParser> s) {
            this.isResourceManaged = isResourceManaged;
            this.supplier = s;
        }
    }
    @ParameterizedTest
    @EnumSource(InputSource.class)
    public void testIonParserIsResourceManaged(InputSource inputSource) throws Throwable {
        assertResourceManaged(inputSource.isResourceManaged, inputSource.supplier);
    }
```

I tried out a few other patterns here but didn't really love any of them. I think non-parameterized is the best for this case. Your mileage may vary.